My Site Dashboard: Phase 2 - Address flicker issue on My Site tab

zwarm opened the pull request:
Fixes #15673  

This PR addresses the flickering issue on the My Site Tab by explicitly comparing contents for the following cards:

- QuickActionsCard
- QuickStartCard
- PostCardsWithPostItems
- PostCardsWithoutPostItems

Before making this adjustment, I added log lines to the `MySiteAdapterDiffCallback` to see where the disconnect was happening. The above cards always returned `false` when `areContentsTheSame` was invoked. 

I thought about adding a `MySiteAdapterDiffCallbackTest` class, but it doesn't look we have any tests (across the app) for testing `areContentsTheSame`. Wdyt?

With this change we shouldn't need to adjust any logic for updating data `onResume` of `MySiteFragment`

**To test:**
Prerequisites
- Launch the app
- Navigate to My Site -> App Settings -> Debug Settings
- Ensure that the `MySiteDashboardPhase2FeatureConfig` is to set to on
- Restart the app if needed

#### Test 1 - PTR
-  Launch the app
- Navigate to the My Site tab
- Swipe down on the view to trigger the refresh
- Note: There is no flickering when the data is refreshed 

#### Test 2 - Switching sites
-  Launch the app
- Navigate to the My Site tab
- Using the site switcher, select a different site 
- Note: The My Site view changes to reflect the newly selected site and there is no flickering. There is still animation to reflect the data has changed for the new site.

#### Test 3 - Return from My Site Tab
-  Launch the app
- Navigate to the My Site tab
- Navigate to another part of the app (Tap a different tab, select a quick action)
- Navigate back to the My Site tab
- Note: There is no flickering 

#### Test 3 - Return from Post update
- Launch the app
- Navigate to a site that has draft post data
- Tap the drafts footer link
- Update the draft post title & save it 
- Navigate back to the My Site tab after the post draft has been uploaded
- Note: The data for the draft post card has changed and there is no flickering. You will still see movement for the updated post.

## Regression Notes
1. Potential unintended areas of impact
My Site Tab flickers and/or cards don't change content

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing

3. What automated tests I added (or what prevented me from doing so) N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

zwarm committed:
Explicitly break up areContentsTheSame for QuickActionsCard, QuickStartCard, PostCardWthPostItems and PostCardWithoutPostItems to remove the flicker on the MySite tab

zwarm committed:
Reformat code block

ParaskP7 reviewed (changes requested):
APPROVE

👋 @zwarm !

I have reviewed and tested this PR as per the instructions, everything works as expected, amazing job in getting to the bottom of this, just this small change and the whole dahsboard feed experience seems to have drastically improved! 🌟 🌟 🌟 
> With this change we shouldn't need to adjust any logic for updating data onResume of MySiteFragment

PS: I actually agree on that. At least for this iteration. If, in the future, we find out that performance is somehow affected and that the dashboard feed experience is degraded, we can invest more time in making sure that the `refresh()` functionality happen explicitly when needed and not implicitly during `onResume()` every single time. For now, the adapter and its diff callback functionality saves us.

-----

However, I am requesting a change because I think it is better if we go another route. Let me explain.

What we are seeing here is a data class problem, where the equality is not working as expected due to lambdas. This is very dangerous and thus we are seeing out adapter and diff callback misbehaving. So, although what you did works, you actually manually implemented the equals operator for every type of card where we have this problem. What I recommend instead is going back to `MySiteCardAndItem` and for example the `PostCard` sealed class and its data subclasses and update the structure so that the lambdas end up not being part of the data class and thus for the equality check to start work as expected.

Also, there is this article I found about using lambdas in data classes and adapter usage, which utilises method reference to point to the same function for every item and that way it ensures that lambda usage doesn't break the equality check (see [here](https://iamjonfry.com/posts/lambdas-in-data-classes/)).

The problem is that, if we are to go with updating the diff callback instead of the data classes, going forward we would need to manual maintain this class every time the underneath models are changed. Especially when new fields are added, then this can get dangerous as the diff callback, if left unchanged, will not update the content of the adapter as it will think that there is no diff.

I have tried that myself and updated the `onClick` lambda structure within the `PostCardWithPostItems`. The result was identical, there was no more flickering. The footer `onClick` seems to work fine with the method reference that is being passed in it from the parent classes. Not sure, why this is not happening with every lambda. 🤔 

Note, that I also strongly recommend that this is now a good time for us to also add tests, on the diff callback and/or the card items sealed/data classes themselves. We need to make sure that this logic remains robust especially as we progress with the dashboard feed and keep adding new cards to it. What I am trying to say is that we now need to guard against this going forward.

Your thoughts?

zwarm committed:
Refactor: restore original areContentsTheSame function

zwarm committed:
Remove TooManyFunctions annotation

zwarm committed:
Change postCard onClick type to ListItemInteraction

zwarm committed:
Refactor: swap lambda for ListItemInteraction.create for the onClick of the PostItem

zwarm committed:
Refactor: swap invoke for ListItemInteraction.onClick

zwarm committed:
Refactor: replace lambda on call to ListItemInteraction.create

zwarm committed:
Refactor: replace lambda call with ListItemInteraction.create

zwarm committed:
Fix failing tested related to onClick in PostItem refactored to ListItemInteraction

zwarm commented:
@ParaskP7 Thanks for the great feedback. 👠  ✋  (that's a high-five - why doesn't that exist here?)

I reworked the following:
- `PostItem.onClick` was changed to a `ListItemInteraction`.  This makes it consistent with all the other "List Item" interactions.
- `QuickStartCardBuilder` and `QuickActionsCardBuilder` removed the lambda syntax on `ListItemInteraction.create`

Using `ListItemInteraction` is not foolproof either, because you can still invoke it like so...
`onClick = ListItemInteraction.create { params.onClick.invoke() }` 
This will also cause the diffUtil `areContentsTheSame` to return false, because it thinks the lambda's aren't the same (Yes, even if it originated from a function pointer). 

To get `areContentTheSame` to work correctly, I used the following syntax:
                `onClick = ListItemInteraction.create(post.id, onPostItemClick)`

Note: This is the syntax used for DomainRegistration ([here](https://github.com/wordpress-mobile/WordPress-Android/blob/9c12166219766c03a1e952fd9cc99870f670ced0/WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/CardsBuilder.kt#L63)), SiteInfoCardBuilder ([here](https://github.com/wordpress-mobile/WordPress-Android/blob/9c12166219766c03a1e952fd9cc99870f670ced0/WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/siteinfo/SiteInfoCardBuilder.kt#L38)) and ListItems ([here](https://github.com/wordpress-mobile/WordPress-Android/blob/9c12166219766c03a1e952fd9cc99870f670ced0/WordPress/src/main/java/org/wordpress/android/ui/mysite/items/SiteItemsBuilder.kt#L33)) etc. (basically what is working correctly) 

There are two warnings in `ListItemInteraction` one [here](https://github.com/wordpress-mobile/WordPress-Android/blob/9c12166219766c03a1e952fd9cc99870f670ced0/WordPress/src/main/java/org/wordpress/android/ui/utils/ListItemInteraction.kt#L7) and one [here](https://github.com/wordpress-mobile/WordPress-Android/blob/9c12166219766c03a1e952fd9cc99870f670ced0/WordPress/src/main/java/org/wordpress/android/ui/utils/ListItemInteraction.kt#L16) 


> Also, there is this article I found about using lambdas in data classes and adapter usage, which utilises method reference to point to the same function for every item and that way it ensures that lambda usage doesn't break the equality check (see [here](https://iamjonfry.com/posts/lambdas-in-data-classes/)).

Isn't this what we are already doing?  Creating the method in `MySiteViewModel` and passing that along to the Params for building?
```
            PostCardBuilderParams(
                    posts = cards?.model?.firstOrNull { it is PostsCardModel } as? PostsCardModel,
                    onPostItemClick = this::onPostItemClick,
                    onFooterLinkClick = this::onPostCardFooterLinkClick
            ),
```

> The problem is that, if we are to go with updating the diff callback instead of the data classes, going forward we would need to manual maintain this class every time the underneath models are changed. Especially when new fields are added, then this can get dangerous as the diff callback, if left unchanged, will not update the content of the adapter as it will think that there is no diff.

I agree 👍 .

> I have tried that myself and updated the `onClick` lambda structure within the `PostCardWithPostItems`. The result was identical, there was no more flickering. The footer `onClick` seems to work fine with the method reference that is being passed in it from the parent classes. Not sure, why this is not happening with every lambda. 🤔

I would love to see what you changed, because I'm pretty sure we didn't implement it the same way. I followed the pattern used in the cards/items that were not triggering the `areContentsTheSame` failure.

> Note, that I also strongly recommend that this is now a good time for us to also add tests, on the diff callback and/or the card items sealed/data classes themselves. We need to make sure that this logic remains robust especially as we progress with the dashboard feed and keep adding new cards to it. What I am trying to say is that we now need to guard against this going forward.

I'm not 100% sure that we can prevent someone from not using a method reference or even using one and then having the syntax correct. I started going down this route to create tests, but stopped because it really isn't clear what we'd be testing. 

Have a new look around. Thanks :)

ParaskP7 reviewed (approved):
👋 @zwarm !

This is amazing, thanks so much for reworking the solution, I review and tested it and everything works like a charm, kudos! ❤️ 🌟 🌟

> I'm not 100% sure that we can prevent someone from not using a method reference or even using one and then having the syntax correct. I started going down this route to create tests, but stopped because it really isn't clear what we'd be testing.

About that, I understand your confusion. What I was thinking is us creating a test of some kind which will construct two exactly same instances with all the available my site adapter items, and then compare them by asserting equality. I am not sure if this test can be somehow be part of the `MySiteAdapterDiffCallback` and thus called from the adapter itself or be as simple as a `MySiteCardAndItem` test. I understand that it might be difficult to enforce usage of `ListItemInteraction`, and more so, a correct usage, since as you showcased in some cases it was incorrectly used. Maybe that might even be a good candidate for a `Lint` rule, not sure.

The only things that I am certain of is that if we don't do something about it, we will end up with the same problem in the future. Maybe we can just create an issue for now, describe all that and add subtasks to try and create tests and lint rules to make that robust from future updates. Wdyt?
